metronome/networking/src/io/iohk/metronome/networking/ConnectionHandler.scala

```diff
@@ -0,0 +1,274 @@
+    connectionQueue: ConcurrentQueue[F, HandledConnection[F, K, M]],
+    connectionsRegister: ConnectionsRegister[F, K, M],
+    messageQueue: ConcurrentQueue[F, MessageReceived[K, M]],
+    cancelToken: TryableDeferred[F, Unit],
+    connectionFinishCallback: HandledConnection[F, K, M] => F[Unit]
+) {
+
+  private val numberOfRunningConnections = AtomicInt(0)
+
+  private def closeAndDeregisterConnection(
+      handledConnection: HandledConnection[F, K, M]
+  ): F[Unit] = {
+    for {
+      _ <- Concurrent[F].delay(numberOfRunningConnections.decrement())
+      _ <- connectionsRegister.deregisterConnection(handledConnection)
+      _ <- handledConnection.close
+    } yield ()
+  }
+
+  /** Registers connections and start handling incoming messages in background, in case connection is already handled
+    * it closes it
+    *
+    * @param possibleNewConnection, possible connection to handle
+    */
+  def registerOrClose(
+      possibleNewConnection: HandledConnection[F, K, M]
+  ): F[Unit] = {
+    connectionsRegister.registerIfAbsent(possibleNewConnection).flatMap {
+      case Some(_) =>
+        //TODO [PM-3092] for now we are closing any new connections in case of conflict, we may investigate other strategies
+        // like keeping old for outgoing and replacing for incoming
+        possibleNewConnection.close
+      case None =>
+        connectionQueue.offer(possibleNewConnection)
+    }
+  }
+
+  /** Checks if handler already handles connection o peer with provided key
+    *
+    * @param connectionKey key of remote peer
+    */
+  def isNewConnection(connectionKey: K): F[Boolean] = {
+    connectionsRegister.isNewConnection(connectionKey)
+  }
+
+  /** Retrieves set of keys of all connected and handled peers
+    */
+  def getAllActiveConnections: F[Set[K]] =
+    connectionsRegister.getAllRegisteredConnections.map { connections =>
+      connections.map(_.key)
+    }
+
+  /** Number of connections actively red in background
+    */
+  def numberOfActiveConnections: F[Int] = {
+    Concurrent[F].delay(numberOfRunningConnections.get())
+  }
+
+  /** Stream of all messages received from all remote peers
+    */
+  def incomingMessages: Iterant[F, MessageReceived[K, M]] =
+    Iterant.repeatEvalF(messageQueue.poll)
+
+  /** Retrieves handled connection if one exists
+    *
+    * @param key, key of remote peer
+    */
+  def getConnection(key: K): F[Option[HandledConnection[F, K, M]]] =
+    connectionsRegister.getConnection(key)
+
+  def sendMessage(
+      recipient: K,
+      message: M
+  ): F[Either[ConnectionAlreadyClosedException[K], Unit]] = {
+    getConnection(recipient).flatMap {
+      case Some(connection) =>
+        connection
+          .sendMessage(message)
+          .attemptNarrow[ConnectionAlreadyClosed]
+          .flatMap {
+            case Left(_) =>
+              connection.close.map { _ =>
+                Left(ConnectionAlreadyClosedException(recipient))
+              }
```

Comment:
Shouldn't this be `closeAndDeregisterConnection(connection).as(Left(...))`?

Reply:
actually, calling `close` on channel will cause `incomingMessage` to generate `None` event which means that our stream handler will fire 
```
closeAndDeregisterConnection(connection).flatMap(_ => callCallBackIfNotClosed(connection))
```
bit.
But here we really depend on how closing is implemented in scalanet (i.e close on channel closes queue in only generates `None` events later), another option would be to have our own `requestClose: Deferred[]` in `HandledConnection`, which would generate `None` event on stream,  and make this closing logic to fire.  wdyt ?

Reply:
Ah, okay, I haven't put this together in my head before. I don't have a problem with that, the incoming returning `None` is a pretty fundamental contract of signalling closure. 

Now I'm looking for where it's put back in the "connections to acquire" queue :sweat_smile:

Reply:
Right, the "finish callback". I think a comment would be enough to explain this flow, I'm sure it's not going to be obvious for people down the line.